### PR TITLE
refactor: build chain extension method

### PR DESCRIPTION
**Pull request by chungquantin**
Link to discussion: https://github.com/r0gue-io/pop-node/pull/71#discussion_r1683692879

Contract size: Original wasm size: 31.2K, Optimized: 7.5K

```
pub(crate) fn build_extension_method(
	version: u8,
	function: u8,
	module: u8,
	dispatchable: u8,
) -> ChainExtensionMethod<(), (), (), false> {
	ChainExtensionMethod::build(u32::from_le_bytes([version, function, module, dispatchable]))
}
```

### CODE REVIEW

**Comment by evilrobot-01** [comment]
An alternative implementation which keeps things simpler imo, taking advantage of scopes: https://github.com/r0gue-io/pop-node/commit/ee4d4994bbef1bb6bc4a587b22e23d517d868677

- reuses existing `lib.rs `rather than introducing a new module for a single function. `build_extension_method` will probably be used in almost every submodule, so easier if its globally available rather than within a separate `utils` module.
- adds a `build_extension_method` at the `v0` module root, so we dont have to repeat V0 everywhere within the module
- adds simple `build_dispatch` and `build_read_state` functions as well, to avoid having to repeat the `DISPATCH` and `READ_STATE` constants throughout all the implemented functions. We could drop the `build_`* prefix on these functions too.
- no `pub(crate)` modifiers required

**Comment by Daanvdplas** [comment]
Could you also add the contract size in the description of the PR with the changes made (with `--release`)

File: pop-api/src/lib.rs [hunk 1]
```diff
@@ -33,6 +33,16 @@ mod constants {
 	pub(crate) const BALANCES: u8 = 10;
 }
 
+/// Helper method to build `ChainExtensionMethod``
```
Comment (evilrobot-01) [inline pop-api/src/lib.rs:36]: Can we add comments for the arguments please? We can then simply copy them to the version specific functions.

File: pop-api/src/lib.rs [hunk 2]
```diff
@@ -33,6 +33,16 @@ mod constants {
...
+pub fn build_extension_method(
```
Comment (evilrobot-01) [inline pop-api/src/lib.rs:37]: Don't expect this to be used outside the crate so pub unnecessary for now. We want devs to use the API functions, hiding this complexity from them.

File: pop-api/src/v0/assets/mod.rs [hunk 3]
```diff
@@ -91,13 +91,11 @@ const TRANSFER_APPROVED: u8 = 25;
 /// Move some assets from the sender account to another, keeping the sender account alive.
 #[inline]
 pub fn transfer_keep_alive(id: AssetId, target: AccountId, amount: Balance) -> Result<()> {
-	ChainExtensionMethod::build(u32::from_le_bytes([
-		V0,
-		DISPATCH,
+	build_dispatch(
 		ASSETS,
 		// E.D. is always respected with transferring tokens via the API.
```
Comment (evilrobot-01) [inline pop-api/src/v0/assets/mod.rs:96]: Be good to place this comment above the build dispatch call, only so it fits on a single line.

File: pop-api/src/v0/assets/mod.rs [hunk 4]
```diff
@@ -222,7 +220,7 @@ pub fn token_symbol(id: AssetId) -> Result<Vec<u8>> {
 
 #[inline]
 pub fn token_decimals(id: AssetId) -> Result<u8> {
-	ChainExtensionMethod::build(u32::from_le_bytes([V0, READ_STATE, ASSETS, TOKEN_DECIMALS]))
+	build_read_state(ASSETS, TOKEN_DECIMALS)
```
Comment (evilrobot-01) [inline pop-api/src/v0/assets/mod.rs:223]: We could have another set of build_dispatch and build_read_state functions defined at the v0::assets module scope to remove the need for the repeated ASSETS constant use.

Essentially providing the same functions at each level, narrowing the scope from general to somewhat specific.

File: pop-api/src/v0/mod.rs [hunk 5]
```diff
@@ -16,3 +22,22 @@ impl From<StatusCode> for Error {
 		value.0.into()
 	}
 }
+
+/// Helper method to build `ChainExtensionMethod``
+pub fn build_extension_method_v0(
```
Comment (evilrobot-01) [inline pop-api/src/v0/mod.rs:27]: As above, can remove pub and add arguments doc comments. Same with the other two functions below.

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
Looking good, think there are a few final possible improvements and then should be good to merge.

File: pop-api/src/lib.rs [hunk 6]
```diff
@@ -34,7 +34,12 @@ mod constants {
 }
 
 /// Helper method to build `ChainExtensionMethod``
-pub fn build_extension_method(
+///
+/// - `version`: The version of the chain extension
+/// - `function`: The ID of the function
+/// - `module`: The index of the runtime module
+/// - `dispatchable`: The index of the module dispatchable functions
+fn build_extension_method(
```
Comment (Daanvdplas) [inline pop-api/src/lib.rs:42]: Please document like:

```
/// <Description>.
///
/// Parameters:
/// - '<parameter>': <description>.
```

File: pop-api/src/v0/mod.rs [hunk 7]
```diff
@@ -33,11 +37,17 @@ pub fn build_extension_method_v0(
...
+/// - `dispatchable`: The index of the module dispatchable functions
+fn build_dispatch(module: u8, dispatchable: u8) -> ChainExtensionMethod<(), (), (), false> {
 	build_extension_method_v0(DISPATCH, module, dispatchable)
 }
 
 /// Helper method to build a dispatch call `ChainExtensionMethod`
-pub fn build_read_state(module: u8, dispatchable: u8) -> ChainExtensionMethod<(), (), (), false> {
+///
+/// - `module`: The index of the runtime module
+/// - `dispatchable`: The index of the module dispatchable functions
```
Comment (Daanvdplas) [inline pop-api/src/v0/mod.rs:50]: This is strictly not a dispatchable but a `state_query`

File: pop-api/src/lib.rs (hunk 6 above)
Comment (Daanvdplas) [reply pop-api/src/lib.rs:42]: This is inconsistent in the repo and will be fixed (#105), my apologies.

Comment (chungquantin) [reply pop-api/src/lib.rs:42]: @Daanvdplas So should we ignore it and make a refractor in the other PR?

Comment (Daanvdplas) [reply pop-api/src/lib.rs:42]: No you can update here already

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
Looking nice and clean! One final improvement suggested.

File: pop-api/src/v0/mod.rs [hunk 8]
```diff
@@ -16,3 +22,35 @@ impl From<StatusCode> for Error {
...
 	}
 }
+
+/// Helper method to build `ChainExtensionMethod` for version `v0`
+///
+/// Parameters:
+/// - 'function': The ID of the function
+/// - 'module': The index of the runtime module
+/// - 'dispatchable': The index of the module dispatchable functions
+fn build_extension_method_v0(
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pop-api/src/v0/mod.rs]: This is resolved. Thanks for the feedback

**Review by evilrobot-01 — APPROVED** [review]
LGTM, thanks for your patience! This will be a very important piece to everything that needs to be built out, so great to have such a tidy implementation.

**Comment by Daanvdplas** [comment]
I think something went wrong when merging 'daan/api' because the helper functions are now not being used in the fungibles apis.

**Comment by chungquantin** [comment]
@Daanvdplas Rebased. Should be ok now. Can you check it again?

**Comment by Daanvdplas** [comment]
Good work @chungquantin!

**Pull request merged**